File: Haskell/FindTheUniqueNumber.hs
{-
---Find the Unique Number---
Write a function called findUnique which returns the only unique 
number from an array.

All numbers in the unsorted array are present twice, except the one 
you have to find. The numbers are always valid integer values between 
1 and 2147483647, so no need for type and error checking. The array contains 
at least one number and may contain millions of numbers. So make sure your 
solution is optimized for speed.

Example
Input:

[ 1, 8, 4, 4, 6, 1, 8 ]
Expected output:

6
-}

import           Data.List

main = print $ findUnique [-59, -5, -5]

findUnique :: [Int] -> Int
findUnique ds | length ds > 1 = discoverUnique (sort ds) 1000000
              | otherwise     = head ds

discoverUnique :: [Int] -> Int -> Int
discoverUnique (d : ds) dp | d `elem` ds || d == dp = discoverUnique ds d
                           | otherwise              = d

File: Python/Find_the_Unique_Number.py
"""
There is an array with some numbers. All numbers are equal except for one. Try to find it!

find_uniq([ 1, 1, 1, 2, 1, 1 ]) == 2
find_uniq([ 0, 0, 0.55, 0, 0 ]) == 0.55
It’s guaranteed that array contains at least 3 numbers.

The tests contain some very huge arrays, so think about performance.
"""

def find_uniq(arr):
    # Make the input list unique
    integers = list(set(arr))
    
    # Filter out all instances in input array for each number in unique array
    integer1_times = list(filter(lambda x: x == integers[0], arr))
    integer2_times = list(filter(lambda x: x == integers[1], arr))
    
    # return the correct unique number 
    return integer1_times[0] if len(integer1_times) < len(integer2_times) else integer2_times[0]